File: Retrofit excercise/app/src/main/java/hr/ferit/davidbilic/retrofitexcercise/MainActivity.java
package hr.ferit.davidbilic.retrofitexcercise;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MainActivity extends AppCompatActivity {

    private Call<List<Product>> apiCall;

    private Button btnSearchProduct;
    private EditText etSearchProduct;

    private RecyclerView recycler;
    private RecyclerAdapter adapter;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        initializeUI();
        setUpRecycler();

        btnSearchProduct.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                setUpAPIcall(etSearchProduct.getText().toString());
            }
        });

    }


    public void initializeUI(){
        btnSearchProduct= findViewById(R.id.btnSearchProduct);
        etSearchProduct= findViewById(R.id.etSearchProduct);
        recycler=findViewById(R.id.rvProducts);
    }

    public void setUpRecycler(){
        recycler.setLayoutManager(new LinearLayoutManager(this));
        adapter= new RecyclerAdapter();
        recycler.setAdapter(adapter);
    }

    private void setUpAPIcall(String brand) {
        apiCall = NetworkUtils.getApiInterface().getProducts(brand);
        apiCall.enqueue(new Callback<List<Product>>() {

            @Override
            public void onResponse(Call<List<Product>> call, Response<List<Product>> response) {
                if (response.isSuccessful() && response.body() != null) {
                    adapter.addProducts(response.body());
                }
                if (response.isSuccessful() && response.body().size()==0) {
                    etSearchProduct.setError("No product matches search request");
                }
            }

            @Override
            public void onFailure(Call<List<Product>> call, Throwable t) {
                Toast.makeText(MainActivity.this, t.getMessage(), Toast.LENGTH_LONG).show();
            }
        });
    }

    @Override
    protected void onDestroy() {
    super.onDestroy();
    if (apiCall != null)
        apiCall.cancel();
}
}

File: KotlinRetrofitExcercise-TvShows/app/src/main/java/gg/bidavid/tvshowsapp/MainActivity.kt
package gg.bidavid.tvshowsapp

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Toast
import androidx.recyclerview.widget.GridLayoutManager
import gg.bidavid.tvshowsapp.model.SearchResult
import gg.bidavid.tvshowsapp.networking.NetworkUtils
import kotlinx.android.synthetic.main.activity_main.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

const val POSITION_KEY: String = "POSITION"
const val RESPONSE_LIST_KEY : String = "LIST"

class MainActivity : AppCompatActivity(), ClickListener {

    lateinit var searchResultList: ArrayList<SearchResult>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setUpRecycler()
        setUpListeners()
    }

    private fun setUpRecycler() {
        recyclerView.layoutManager= GridLayoutManager(this, 4)
        recyclerView.adapter= RecyclerAdapter(this)
    }

    private fun setUpListeners() {
        btnSearch.setOnClickListener{findShows()}
    }

    private fun findShows() {
        val title = etTVshowName.text.toString().trim()

        NetworkUtils.apiInterface.getTvShows(title).enqueue(object: Callback<List<SearchResult>>{
                    override fun onFailure(call: Call<List<SearchResult>>, t: Throwable) {
                        showToast(t.message)
                    }

                    override fun onResponse(call: Call<List<SearchResult>>, response: Response<List<SearchResult>>) {
                       if(response.isSuccessful && response.body()!= null){
                           showToast("Successfull")
                           searchResultList=ArrayList(response.body()!!)
                           setUpRecyclerData(searchResultList)
                       }
                    }

                })

    }

    private fun showToast(message: String?) {
        Toast.makeText(this, message, Toast.LENGTH_LONG).show()
    }

    private fun setUpRecyclerData(retrievedShows: List<SearchResult>) {
        (recyclerView.adapter as RecyclerAdapter).addData(retrievedShows)
    }

    override fun OnPosterClick(position: Int) {
        Toast.makeText(this, "$position",Toast.LENGTH_SHORT).show()
        val intent = Intent(this, TvShowDetailsActivity::class.java)
        intent.putExtra(POSITION_KEY, position)
        intent.putExtra(RESPONSE_LIST_KEY, searchResultList) //RADI OVOG SE RUSI, jer nisam implementirao serializable u klase
        startActivity(intent)
    }

}
